### PR TITLE
remove tupling for foreach in keySet/values of immutable HashMap and TreeMap

### PR DESCRIPTION
**Pull request by mkeskells**

### CODE REVIEW

**Comment by hrhino** [comment]
MiMa complaints seem ignorable.

File: src/library/scala/collection/immutable/HashMap.scala [hunk 1]
```diff
@@ -139,6 +139,21 @@ sealed class HashMap[A, +B] extends AbstractMap[A, B]
...
+  /* Override to avoid tuple allocation in foreach */
+  private[collection] class HashMapKeys extends ImmutableDefaultKeySet {
+    override def foreach[U](f: A => U) = foreachEntry((key, _) => f(key))
+    override lazy val hashCode = super.hashCode()
+  }
+  override def keySet: immutable.Set[A] = new HashMapKeys
+
+  /** The implementation class of the iterable returned by `values`.
+   */
+  private[collection] class HashMapValues extends DefaultValuesIterable {
```
Comment (dwijnand) [inline src/library/scala/collection/immutable/HashMap.scala:151]: > MiMa complaints seem ignorable.

Could these classes be private?  They can't escape into the API, for binary compatibility, anyways.

Comment (mkeskells) [reply src/library/scala/collection/immutable/HashMap.scala:151]: I would like to be able to use them later for fast pathing the hashset operations like ++ as they share the same structure. We could make then private now though

Comment (dwijnand) [reply src/library/scala/collection/immutable/HashMap.scala:151]: Ah, right.  Half of my motivation was wondering/hoping it removed the MiMa false positive, anyways.

**Comment by szeiger** [comment]
How does this avoid tupling when all calls still go through `foreachEntry`?

**Comment by mkeskells** [comment]
Foreachentry is a function that takes 2 parameters not a tuple2. This means that the mapoperatiisns like hashcode dont create tuplea to do the hashcode

**Review by szeiger — APPROVED** [review]

**Pull request merged**